Added StringBehaviorExpectation and tests

scpeterson opened the pull request:
This is my first ever open source PR, so please go easy on me...  This is my attempt to add the `EmptyStringBehaviorExpectation` portion of Issue #896.  It appeared to me that the new `EmptyStringBehaviorExpectation` class was very similar to the `EmptyGuidBehaviorExpectation`class, so it will look similar.  I added it to `GuardClauseAssertion` as well, uncertain if there was a set goal to make this truly optional.  I appreciate any feedback.

scpeterson commented on Src/Idioms/EmptyStringBehaviorExpectation.cs:
```diff
@@ -0,0 +1,49 @@
+            }
+            catch (Exception e)
+            {
+                throw command.CreateException("\"string.Empty\"", e);
```
In this case, I was trying to be consistent with how things were done in `EmptyGuidBehaviorExpectation`  but I see your point and can make the change.

scpeterson commented on Src/Idioms/GuardClauseAssertion.cs:
```diff
@@ -32,7 +32,8 @@ public class GuardClauseAssertion : IdiomaticAssertion
                 new NullReferenceBehaviorExpectation(),
-                new EmptyGuidBehaviorExpectation()))
+                new EmptyGuidBehaviorExpectation(),
+                new EmptyStringBehaviorExpectation()))
```
This was my main area of concern; I will remove it so I don't break things.

scpeterson commented on Src/IdiomsUnitTest/EmptyStringBehaviorExpectationTest.cs:
```diff
@@ -0,0 +1,134 @@
+        public void VerifyCorrectlyInvokesExecuteWhenRequestedTypeIsString()
+        {
+            // Arrange
+            var mockVerified = false;
```
This too is consistent with `EmptyGuidBehaviorExpectation` but I will make this change as well.

scpeterson commented on Src/IdiomsUnitTest/GuardClauseAssertionTest.cs:
```diff
@@ -1727,6 +1728,11 @@ public void Method(string argument1, out object argument2, int argument3)
                     throw new ArgumentNullException(nameof(argument1));
                 }
 
+                if (argument1 == string.Empty)
```
I believe you are correct; once I remove my breaking change, I won't need the change I made here.

scpeterson replied:
Do we still need the integration tests if I removed the new empty string guard from default behavior?

zvirja replied:
Yes, we definitely need.

Currently we have two kind of tests: unit and integration (or facade). Previously you have added a set of unit tests in the `EmptyStringBehaviorExpectationTest` file to verify that it behaves correctly. However, it would be cool to also have tests where we initialize real guard assertion with our custom behavior and verify the real methods. While tests overlap a bit, I believe that we still have a lot of benefits from it.

scpeterson commented on Src/IdiomsUnitTest/EmptyStringBehaviorExpectationTest.cs:
```diff
@@ -0,0 +1,134 @@
+            {
+                sut.Verify(mockCommand);
+            }
+            catch (InvalidOperationException) { }
```
This too is how it was done elsewhere.  I'll look at it a bit and see if I can't make it more specific.

scpeterson commented on Src/Idioms/EmptyStringBehaviorExpectation.cs:
```diff
@@ -0,0 +1,49 @@
+            {
+                command.Execute(string.Empty);
+            }
+            catch (ArgumentException)
```
`NullReferenceBehaviorExpectation` uses `ArgumentNullException` which takes a parameter name in its constructor, whereas `ArgumentException` does not.  I believe this is why `EmptyGuidBehaviorExpectation` does not have the same parameter name test as `NullReferenceBehaviorExpectation`.  I could test for a message, like "The parameter name is invalid" or something like that.

zvirja replied:
I've just tested that and found that can pass parameter name to the `ArgumentException`:
```c#
var exception = new ArgumentException("Empty string", nameof(arg));
var paramName = exception.ParamName;
```

Could you please verify for one more time whether I'm missing something?

scpeterson replied:
Clearly, my brain was operating poorly...  I made the change and added the test.

scpeterson replied:
However...  `ArgumentNullException` has an overload that allows you to call it with just a parameter name; `ArgumentException` does not.  Calling `ArgumentException` with only one parameter causes `message` to be populated, not `parameterName`.  This is problematic if a person doesn't create their Guard Clause for empty strings with both a message and a parameter name.  `EmptyGuidBehaviorExpectation` doesn't have this problem since it does not have the same parameter check as `NullReferenceBehaviorExpectation`.  In other words, with the way I wrote `EmptyStringBehaviorExpectation`, this Guard Clause in a class being verified will not work:

        private class EmptyStringTest
        {
            public EmptyStringTest(string bogus)
            {
                if (bogus == string.Empty)
                    throw new ArgumentException(nameof(bogus));
            }
        }
However, this will work for `EmptyGuidBehaviorExpectation`:

        private class EmptyGuidTest
        {
            public EmptyGuidTest(Guid bogus)
            {
                if (bogus == Guid.Empty)
                    throw new ArgumentException(nameof(bogus));
            }
        }

They look the same, but, since `EmptyGuidBehaviorExpectation` does NOT compare parameter names, it works if a Guid is empty, whereas `EmptyStringBehaviorExpectation` does not work if a string is empty since the parameter name is null.  To make my current implementation of `EmptyStringBehaviorExpectation` function properly, a developer would have to construct their Guard Clause (for empty strings) like so in order to populate the parameter name:

        private class EmptyStringTest
        {
            public EmptyStringTest(string bogus)
            {
                if (bogus == string.Empty)
                    throw new ArgumentException("my message", nameof(bogus));
            }
        }

So to me, it feels like the expected behavior between empty Guid and empty String shouldn't be different; I should probably remove the parameter check.  The other option is to add a parameter check to empty Guid, but that would be a breaking change.

Any thoughts?

zvirja replied:
Thanks for describing your point in detail! I've reviewed it and here are my thoughts.

Personally, I feel that `EmptyGuidBehaviorExpectation` should also verify the parameter name. I believe that the reason why it doesn't do this is that we simply overlooked the inconsistency. That's something that should be fixed in v5 and [I've created](https://github.com/AutoFixture/AutoFixture/issues/1009) an issue for that.

My reasoning is that if you have the following guard (from the sample above):
```c#
if (string.IsNullOrEmpty(secondArg)) throw new ArgumentException(nameof(secondArg));
```
then the thrown exception doesn't contain any explanation of what happened:
```
System.ArgumentException
secondArg
```

That's not what you expect from the exception message as it should contain the meaningful message. Compare that to `ArgumentNullException`:
```
System.ArgumentNullException
Value cannot be null.
Parameter name: secondName
```

For me, the goal of the `GuardClauseAssertion` is to:
- verify a guard clause is present
- verify a guard clause is valid, i.e. it doesn't violate the common .NET contract.

I feel that exception without meaningful message definitely violates the .NET contract and we should help to catch such things.

Notice, you have a few ways to have nice exceptions and don't duplicate message all over the code:
- Create aux method to verify value (e.g. `Assert.ArgumentIsNotNullOrEmpty(..)`), so message is specified there.
-  Create a custom exception like `ArgumentNullException` taking the parameter name only.

Actually, the last option is preferable to me and here is how it looks:
```c#
public class ArgumentNullOrEmptyException : ArgumentException
{
    public ArgumentNullOrEmptyException(string paramName)
        : base("Value should not be null or empty.", paramName)
    {
    }
}

......
if (string.IsNullOrEmpty(secondArg)) throw new ArgumentNullOrEmptyException(nameof(secondArg));
```

Result:
```
AutoFixturePlayground.ArgumentNullOrEmptyException
Value should not be null or empty.
Parameter name: secondArg
```

A few minor changes to the code and now your log is not full of awkward errors - isn't that awesome? :)

zvirja commented:
@scpeterson Please take a look at the registered #1010. May I ask you ensure that the introduced `EmptyStringBehaviorExpectation` doesn't suffer from the similar issue?

I see the following way:
1. Extend the `IGuardClauseCommand` interface with `CreateExceptionInvalidGuard(string explanation, Exception ex)`: 
`An attempt was made to XXXXXXXXX, and Guard Clause prevented this. However the captured Guard Clause is invalid: {0}"`
2. Call this method if `ArgumentException` is detected, however argument name doesn't match: 
```
the thrown exception doesn't contain the valid parameter name.
Expected parameter name: "", actual name: "".

Notice, you should pass parameter name as the _second_ argument of the `ArgumentException` constructor: throw new ArgumentException("message", "argName").
```

If you find how the message might be improved - feel free to enhance it.

scpeterson commented:
I added `ArgumentNullOrEmptyException` and associated tests.  I also updated `EmptyStringBehaviorExpectation` to check for an invalid parameter and make use of the new exception.  I have not yet verified that the #1010 issue has been resolved, but will do that next when time allows.

zvirja commented on Src/Idioms/ArgumentNullOrEmptyException.cs:
```diff
@@ -0,0 +1,64 @@
+    /// Represents an error about a null or empty string.
+    /// </summary>
+    [Serializable]
+    public class ArgumentNullOrEmptyException : ArgumentException
```
Sorry for the confusion I caused by my comment, it seems I wasn't clear enough 😖 

We should not introduce this exception in our library. By doing so we'll force people to reference Idioms from the production code - this wrong a bit, as library was designed to be used in unit test project only. Instead, we should expect the `ArgumentException` raised with the correct parameters.

I mentioned `ArgumentNullOrEmptyException` to demonstrate that this kind of exception _might be created by users_ in their production code to reduce the boilerplate code.

Please remove the exception and expect the `ArgumentException` in the expectation you introduced.

zvirja commented on Src/IdiomsUnitTest/GuardClauseAssertionTest.cs:
```diff
@@ -1838,5 +1838,44 @@ private abstract class AbstractTypeWithAbstractMethod
+        private class ClassWihoutEmptyStringGuard
+        {
+            public ClassWihoutEmptyStringGuard(string param)
+            {
```
Please add null guard here to make the test more close to reality :wink:

zvirja commented on Src/IdiomsUnitTest/GuardClauseAssertionTest.cs:
```diff
@@ -1838,5 +1838,44 @@ private abstract class AbstractTypeWithAbstractMethod
+        {
+            public ClassWithEmptyStringGuard(string param)
+            {
+                if (param == string.Empty)
```
Please add null guard here to make the test more close to reality :wink:

zvirja commented:
@scpeterson I've just merged #1024, where I covered the part related to the situation when thrown exception contains invalid parameter name. It should significantly simplify this PR. Please rebase your branch on top of the latest `master` to consume those changes.

Do you know whether you can have time in the nearby future to continue your work on this?

Thanks.

scpeterson commented:
Alex:

My hope is to get back into it.  I've been occupied with a job search so
I've backed off a bit.  Hopefully soon.

On Wed, Apr 25, 2018 at 4:36 AM, Alex Povar <notifications@github.com>
wrote:

> @scpeterson <https://github.com/scpeterson> I've just merged #1024
> <https://github.com/AutoFixture/AutoFixture/pull/1024>, where I covered
> the part related to the situation when thrown exception contains invalid
> parameter name. It should significantly simplify this PR. Please rebase
> your branch on top of the latest master to consume those changes.
>
> Do you know whether you can have time in the nearby future to continue
> your work on this?
>
> Thanks.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/AutoFixture/AutoFixture/pull/1005#issuecomment-384206988>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABPaut8jGc1fXm6IzEQw_b5eMQ7rn_PZks5tsDWdgaJpZM4S9Yjt>
> .
>

zvirja commented:
@scpeterson Thanks for the update! Wish you luck then in your searches and be waiting for the updates 😉

zvirja commented:
Closing in favor of #1168
